mountpoint-s3-client/tests/network_interface_config.rs

```diff
@@ -0,0 +1,125 @@
+#![cfg(feature = "s3_tests")]
+
+pub mod common;
+
+use test_case::test_case;
+
+use common::*;
+use mountpoint_s3_client::config::{EndpointConfig, S3ClientConfig};
+use mountpoint_s3_client::error::{HeadObjectError, ObjectClientError::ClientError, ObjectClientError::ServiceError};
+use mountpoint_s3_client::{ObjectClient, S3CrtClient};
+
+#[tokio::test]
+async fn test_empty_list() {
+    let (bucket, prefix) = get_test_bucket_and_prefix("test_empty_list");
+    let key = format!("{prefix}/no-such-key");
+
+    let interface_names = Vec::new();
+    let config = S3ClientConfig::new()
+        .endpoint_config(EndpointConfig::new(&get_test_region()))
+        .network_interface_names(interface_names);
+    let client = S3CrtClient::new(config).expect("client should create OK");
+
+    let err = client
+        .head_object(&bucket, &key)
+        .await
+        .expect_err("head_object should fail as the key doesn't exist");
+    assert!(
+        matches!(err, ServiceError(HeadObjectError::NotFound)),
+        "HeadObject should 404 successfully"
+    );
+}
+
+#[tokio::test]
+async fn test_one_interface_ok() {
+    let (bucket, prefix) = get_test_bucket_and_prefix("test_empty_list");
+    let key = format!("{prefix}/no-such-key");
+
+    let primary_interface = get_primary_interface_name();
+    let interface_names = vec![primary_interface];
+    let config = S3ClientConfig::new()
+        .endpoint_config(EndpointConfig::new(&get_test_region()))
+        .network_interface_names(interface_names);
+    let client = S3CrtClient::new(config).expect("client should create OK");
+
+    let err = client
+        .head_object(&bucket, &key)
+        .await
+        .expect_err("head_object should fail as the key doesn't exist");
+    assert!(
+        matches!(err, ServiceError(HeadObjectError::NotFound)),
+        "HeadObject should 404 successfully"
+    );
+}
+
+/// This test demonstrates how the S3 client will fail today when configured with bad network interfaces.
+/// The behavior isn't great, but this documents what happens.
+///
+/// TODO: How can we get to a point where S3 client creation fails when invalid interface names are provided?
```

Comment:
it would be possible if we could affect somehow the interface which is used for the request; then we could've verify validity of all provided interfaces at least on the initial `ListObjectsV2`, may be we can leverage that it's round-robin and do the initial `ListObjectsV2` multiple times defined by NiCs number? 

how does the error provided to `mount-s3` output / logs looks like, when there is a typo in interface's name?

Reply:
>how does the error provided to mount-s3 output / logs looks like, when there is a typo in interface's name?

```
Error: Failed to create S3 client

Caused by:
    0: initial ListObjectsV2 failed for bucket djonesoa-dub in region us-east-1
    1: Client error
    2: Unknown CRT error
    3: CRT error 1053: aws-c-io: AWS_IO_SOCKET_INVALID_OPTIONS, Invalid socket options.
```

You get something similar when it fails at file operation time (such as where one of the interfaces did exist) along with EIO.

Reply:
>may be we can leverage that it's round-robin and do the initial ListObjectsV2 multiple times defined by NiCs number?

I think that could work. I was hoping not to mess too much with that right now since I'd prefer to implement this at the CRT level.

My thinking is that CRT could eagerly do the interface name to ID lookup, and propagate the bad network names immediately at client initialization.